### ISSUE TITLE
CDK FocusMonitor emits invalid state in IE11

### ISSUE THREAD

**Issue by jpzwarte** [angular/components#15517]
#### What is the expected behavior?
In all other browsers when i monitor an element via:
```typescript
  ngAfterViewInit() {
    this._focusMonitor
      .monitor(this.elementRef, true)
//      .pipe(debounceTime(0))
      .subscribe(focusOrigin => {
        if (!focusOrigin && this.radioGroup) {
          this.radioGroup._touch();
        }
      });
  }
```

And i click on the element, 2 values get emitted with `focusOrigin === 'mouse'`.

The workaround i've found is to add a `debounceTime(0)` like the commented line above.

The code in question is from `radio.ts` in material.

#### What is the current behavior?
On IE11, between the `'mouse'` values, i get a `null` value emitted.

#### What are the steps to reproduce?
Try running the example https://stackblitz.com/angular/kgnogyvppxp?file=app%2Fform-field-label-example.ts on IE11 and a console.log in `radio.ts` on the `focusOrigin`.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Latest material, IE11.

**Comment by josephperrott** [comment]
@crisbeto Could you look at this one since you have an IE11 setup already available easily for you?

**Comment by davidjgoss** [comment]
I've found the source of this issue.

When you monitor a parent element with `checkChildren` set to `true`, internally the focus monitor listens for both `focus` and `blur` events on the parent element. When focus shifts from child element 1 to child element 2, the `blur` event on child element 1 is captured first. The design is that this `blur` event does not result in a `null` value on the subject, because the focus monitor will see that the [event's `relatedTarget`](https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/relatedTarget) (the element that's about to take the focus) is also a child of our parent element, [and take no further action](https://github.com/angular/components/blob/2d976a92429ed37a89bfa9ee74b479b48c7ac40d/src/cdk/a11y/focus-monitor/focus-monitor.ts#L375).

However, IE does not set the `relatedTarget` on `focus` or `blur` events, so the focus monitor can't know that focus is not leaving the parent element, and so we get a `null` value on the subject.

For some use cases, we get away with this; if we're just using the `cdkMonitorSubtreeFocus` directive and doing some styling based on `cdk-focused-` classes, no one would ever notice because the offending `blur` event is immediately followed by a `focus` event on child element 2, which "restores" focus to the parent element in the eyes of the focus monitor, so the state is corrected before anything different can even be rendered. Where we are doing destructive stuff programatically based on observing a `FocusOrigin` of `null`, it bites us.

From what I've found, there are two potential fixes:

- Internalise the `.pipe(debounceTime(0))` workaround above, so we give the event loop a chance to catch up before reacting to blur/focus events. This feels like it would be quite a large refactor, and I'm not convinced it would be 100% reliable - depending on how much else is going on in the event loop, it could theoretically be 1ms or more between the events
- Switch to using `focusin` and `focusout` events instead of `focus` and `blur` - for whatever reason, IE _does_ set `relatedTarget` correctly on these, as do the other browsers. I need to do some more research and experimentation to understand the implications of changing this

Comparison of `relatedTarget` across events: https://jsbin.com/cerarinuxu/1/edit?js,console,output
